File: 프로그래머스/lv2/43165. 타겟 넘버/타겟 넘버.cpp
#include <string>
#include <vector>

using namespace std;
int answer = 0;
void dfs(vector<int> numbers, int target, int total, int index){
    if (index == numbers.size()){
        if (total == target){
            answer ++ ;
        }
        return;
    }
    
    dfs(numbers, target, total + numbers[index], index + 1);
    dfs(numbers, target, total - numbers[index], index + 1);
}
int solution(vector<int> numbers, int target) {
    dfs(numbers, target, numbers[0], 1);
    dfs(numbers, target, -numbers[0], 1);
    return answer;
}

File: 프로그래머스/lv2/43165. 타겟 넘버/타겟 넘버.py
from collections import deque

def solution(numbers, target):
    answer = 0
    q = deque()
    q.append((numbers[0], 0))
    q.append((-numbers[0],0))
    while q:
        num, cnt = q.popleft()
        cnt += 1
        if cnt <len(numbers):
            q.append((num+numbers[cnt],cnt))
            q.append((num-numbers[cnt],cnt))
        else:
            if num == target:
                answer += 1
    return answer